opm/material/fluidsystems/GenericFluidSystem.hh

```diff
@@ -0,0 +1,215 @@
+        static const int numMisciblePhases=2;
+        static const int numMiscibleComponents = 3;
+        // TODO: phase location should be more general
+        static constexpr int oilPhaseIdx = 0;
+        static constexpr int gasPhaseIdx = 1;
+
+        template <class ValueType>
+        using ParameterCache = Opm::PTFlashParameterCache<ValueType, GenericFluidSystem<Scalar, NumComp>>;
+        using ViscosityModel = typename Opm::ViscosityModels<Scalar, GenericFluidSystem<Scalar, NumComp>>;
+        using PengRobinsonMixture = typename Opm::PengRobinsonMixture<Scalar, GenericFluidSystem<Scalar, NumComp>>;
+
+        template<typename Param>
+        static void addComponent(const Param& param)
+        {
+            assert(component_param_.size() <= numComponents);
+            if (component_param_.size() == numComponents) {
+                std::cout << " the fluid system has reached maximum " << numComponents << " component,"
+                          << " the component " << param.name << " will not be added " << std::endl;
+            }
+            component_param_.push_back(param);
+        }
+
+        static void init()
+        {
+            component_param_.reserve(numComponents);
+        }
+
+        /*!
+         * \brief The acentric factor of a component [].
+         *
+         * \copydetails Doxygen::compIdxParam
+         */
+        static Scalar acentricFactor(unsigned compIdx)
+        {
+            return component_param_[compIdx].acentric_factor;
+        }
+        /*!
+         * \brief Critical temperature of a component [K].
+         *
+         * \copydetails Doxygen::compIdxParam
+         */
+        static Scalar criticalTemperature(unsigned compIdx)
+        {
+            return component_param_[compIdx].critic_temp;
+        }
+        /*!
+         * \brief Critical pressure of a component [Pa].
+         *
+         * \copydetails Doxygen::compIdxParam
+         */
+        static Scalar criticalPressure(unsigned compIdx)
+        {
+            return component_param_[compIdx].critic_pres;
+        }
+        /*!
+        * \brief Critical volume of a component [m3].
+        *
+        * \copydetails Doxygen::compIdxParam
+        */
+        static Scalar criticalVolume(unsigned compIdx)
+        {
+            return component_param_[compIdx].critic_vol;
+        }
+
+        //! \copydoc BaseFluidSystem::molarMass
+        static Scalar molarMass(unsigned compIdx)
+        {
+            return component_param_[compIdx].molar_mass;
+        }
+
+        /*!
+         * \brief Returns the interaction coefficient for two components.
+         *.
+         */
+        static Scalar interactionCoefficient(unsigned /*comp1Idx*/, unsigned /*comp2Idx*/)
+        {
+            // TODO: some data structure is needed to support this
+            return 0.0;
+        }
+
+        //! \copydoc BaseFluidSystem::phaseName
+        static const char* phaseName(unsigned phaseIdx)
```

Comment:
Why `const char*`?  These all refer to objects of persistent lifetime so could return `std::string_view` instead.

Reply:
We have some old code to deal with when coming to this. 
```c++
                snprintf(name, 512, "gradP_%s", FluidSystem::phaseName(phaseIdx));

```

Any suggestions?

Reply:
Then we change all the `FluidSystem`s. How about we address this by all in a separate PR?

Reply:
I changed to use `std::string_view` for all the `FluidSystem`s. Please let me know if you want me to split this action to a separate PR.